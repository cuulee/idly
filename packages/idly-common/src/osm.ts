export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export type ParentWays = Map<NodeId, Set<WayId>>;
// Table used to map id -> entity
export type EntityTable = Map<EntityId, Entity>;

export type Tags = Map<string, string>;

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export type FeatureProps = { [key: string]: string };

export type NodeGeometry =
  | OsmGeometry.POINT
  | OsmGeometry.VERTEX
  | OsmGeometry.VERTEX_SHARED;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  VERTEX_SHARED = 'vertex_shared', // the one shared among multiple ways
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation'
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: EntityId[];
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: RelationMember[];
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: number;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

export interface RelationMember {
  readonly id: string;
  readonly type: string;
  readonly role: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
