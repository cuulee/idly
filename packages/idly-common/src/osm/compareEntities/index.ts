export { compareNode } from './compareNode';
export { compareRelation } from './compareRelation';
export { compareTags } from './compareTags';
export { compareWay } from './compareWay';
