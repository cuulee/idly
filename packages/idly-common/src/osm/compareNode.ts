import { compareShallow } from './compareShallow';
import { Node, Way } from './structures';

export function compareNode(node: Node, compareWith: Node): boolean {
  if (node === compareWith) {
    return true;
  }

  if (node.id !== compareWith.id) {
    return false;
  }

  if (
    !Object.is(node.loc.lat, compareWith.loc.lat) ||
    !Object.is(node.loc.lon, compareWith.loc.lon)
  ) {
    return false;
  }

  if (!compareShallow(node.tags, compareWith.tags)) {
    return false;
  }

  if (!compareShallow(node.attributes, compareWith.attributes)) {
    return false;
  }

  if (node.type !== compareWith.type) {
    return false;
  }

  return true;
}
