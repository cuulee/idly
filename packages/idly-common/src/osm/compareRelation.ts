import { compareShallow } from './compareShallow';
import { Relation } from './structures';

// readonly id: EntityId;
// readonly type: EntityType.RELATION;
// readonly tags: Tags;
// readonly members: ReadonlyArray<RelationMember>;
// readonly attributes: Attributes;
export function compareRelation(
  relation: Relation,
  compareWith: Relation,
): boolean {
  if (relation === compareWith) {
    return true;
  }

  if (relation.id !== compareWith.id) {
    return false;
  }

  if (
    JSON.stringify(relation.members) !== JSON.stringify(compareWith.members)
  ) {
    return false;
  }

  if (!compareShallow(relation.tags, compareWith.tags)) {
    return false;
  }
  if (!compareShallow(relation.attributes, compareWith.attributes)) {
    return false;
  }

  if (relation.type !== compareWith.type) {
    return false;
  }

  return true;
}
