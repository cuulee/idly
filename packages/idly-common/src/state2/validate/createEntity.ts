import { Either } from 'monet';
import { nodeFactory } from '../../osm/entityFactory/nodeFactory';
import { relationFactory } from '../../osm/entityFactory/relationFactory';
import { wayFactory } from '../../osm/entityFactory/wayFactory';
import {
  Attributes,
  Entity,
  EntityId,
  EntityType,
  LngLat,
  Node,
  Relation,
  RelationMember,
  Tags,
  Way,
} from '../../osm/structures';
import { Err, ErrorWhen } from '../helper';
import {
  checkValidNode,
  checkValidRelation,
  checkValidWay,
} from '../validation';

export interface WayLike {
  readonly id: EntityId;
  readonly type?: 'way';
  readonly tags?: Tags;
  readonly attributes?: Attributes;
  readonly nodes?: EntityId[] | ReadonlyArray<EntityId>;
}
export interface RelLike {
  readonly id: EntityId;
  readonly type?: 'relation';
  readonly tags?: Tags;
  readonly members?: RelationMember[] | ReadonlyArray<RelationMember>;
  readonly attributes?: Attributes;
}

export interface NodeLike {
  readonly id: EntityId;
  readonly type?: 'node';
  readonly tags?: Tags;
  readonly loc?: LngLat;
  readonly attributes?: Attributes;
}
export type EntityLike = NodeLike | WayLike | RelLike;

export const parseIfString = (r: EntityLike | string) =>
  typeof r === 'string' ? (JSON.parse(r) as EntityLike) : r;

export function createEntity(
  entityLike: EntityLike | string,
  type?: EntityType
): Entity {
  return Err.right(entityLike)
    .map(parseIfString)
    .flatMap(
      ErrorWhen(
        r => !!type && type !== r.type,
        new Error('Entity must be ' + type)
      )
    )
    .flatMap(
      ErrorWhen(r => !r.id, new Error('Entity does not have a valid id'))
    )
    .flatMap(entityFrom)
    .cata(err => {
      throw err;
    }, r => r);
}

function entityFrom(entity: any): Either<Error, Entity> {
  switch (entity.type) {
    case EntityType.NODE: {
      return checkValidNode(entity)
        ? Err.right(nodeFactory(entity))
        : Err.left(new Error('Invalid node'));
    }

    case EntityType.WAY: {
      return checkValidWay(entity)
        ? Err.right(wayFactory(entity))
        : Err.left(new Error('Invalid Way'));
    }

    case EntityType.RELATION: {
      return checkValidRelation(entity)
        ? Err.right(relationFactory(entity))
        : Err.left(new Error('Invalid Relation'));
    }
    default:
      return Err.left(new Error('Entity type is unknown'));
  }
}

export const createNode = (entityLike: NodeLike): Node =>
  createEntity({ ...entityLike }, EntityType.NODE) as Node;

export const createWay = (entityLike: WayLike): Way =>
  createEntity({ ...entityLike }, EntityType.WAY) as Way;

export const createRelation = (entityLike: RelLike): Relation =>
  createEntity({ ...entityLike }, EntityType.RELATION) as Relation;
