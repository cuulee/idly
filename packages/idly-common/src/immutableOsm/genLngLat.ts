import { recordify } from "typed-immutable-record";
import { $LngLat } from "../immutableOsm/immutableOsm";
import { LngLat } from "../osm/structures";

export function $genLngLat(l: LngLat): $LngLat {
  return recordify<LngLat, $LngLat>(l);
}
