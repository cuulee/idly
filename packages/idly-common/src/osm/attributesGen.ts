import { Attributes } from '../osm/structures';
const attr = {};

export function attributesGen(
  {
    visible,
    version,
    timestamp,
    changeset,
    uid,
    user
  }: {
    visible?: boolean;
    version?: string;
    timestamp?: string;
    changeset?: string;
    uid?: string;
    user?: string;
  } = {}
): Attributes {
  return {
    changeset,
    timestamp,
    uid,
    user,
    version,
    visible
  };
}
