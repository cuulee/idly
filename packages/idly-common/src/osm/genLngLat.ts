import { LngLat } from "../osm/structures";

export function genLngLat(
  obj: { lon: number; lat: number } | [number, number],
): LngLat {
  if (Array.isArray(obj)) {
    return { lon: obj[0], lat: obj[1] };
  }
  return { ...obj };
}
