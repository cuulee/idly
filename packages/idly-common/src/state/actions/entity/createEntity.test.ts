// tslint:disable no-expression-statement

import { genLngLat } from '../../../osm/genLngLat';
import { createNode } from './createEntity';

// tslint:disable:object-literal-sort-keys

describe('create simple node', () => {
  test('simple test', () => {
    const n = createNode({ id: 'n1' });
    expect(n).toMatchSnapshot();
  });
  test('works with empty tags', () => {
    const n = createNode({ id: 'n1', tags: {} });
    expect(n.tags).toEqual({});
    const n2 = createNode({ id: 'n1' });
    expect(n2.tags).toEqual({});
  });
  test('simple test node with tags', () => {
    const n = createNode({ id: 'n1', tags: { k: 'na' } });
    expect(n).toMatchSnapshot();
  });
  test('works with full blown object', () => {
    const n = createNode({
      id: 'n-a',
      loc: genLngLat([1, 2]),
      tags: { k: 'n-a4' },
      type: 'node',
    });
    expect(n).toMatchSnapshot();
  });
  test('different order of keys produce same output', () => {
    const n1 = createNode({
      loc: genLngLat([1, 2]),
      id: 'n-a',
      type: 'node',
      tags: { k: 'n-a4' },
    });
    const n2 = createNode({
      id: 'n-a',
      loc: genLngLat([1, 2]),
      tags: { k: 'n-a4' },
      type: 'node',
    });
    expect(n2).toEqual(n1);
  });
  test('tags get sorted', () => {
    const n = createNode({ id: 'n1', tags: { k: 'na', a: 'b' } });
    expect(Object.keys(n.tags)).toEqual(['a', 'k']);
  });

  test('works with attributes', () => {
    const n = createNode({ id: 'n1' });
    expect(n.attributes).toEqual({
      changeset: undefined,
      timestamp: undefined,
      uid: undefined,
      user: undefined,
      version: undefined,
      visible: undefined,
    });
  });
  test('works with attributes 2', () => {
    const n = createNode({
      attributes: { visible: false, timestamp: '1' },
      id: 'n1',
    });
    const n2 = createNode({
      id: 'n1',
      attributes: { visible: false, timestamp: '2' },
    });
    expect(n.attributes).not.toEqual(n2.attributes);
  });
  test('works with attributes with differnt key order', () => {
    const n = createNode({
      attributes: { visible: false, timestamp: '1', uid: '1' },
      id: 'n1',
    });
    expect(Object.keys(n.attributes)).toEqual([
      'changeset',
      'timestamp',
      'uid',
      'user',
      'version',
      'visible',
    ]);
  });

  test('reuses already existing instance', () => {
    const n1 = createNode({ id: 'n1', tags: { k: 'na', a: 'b' } });
    const n2 = createNode({ id: 'n1', tags: { k: 'na', a: 'b' } });
    expect(n1).toBe(n2);
  });

  test('reuses already existing instance with reordered keys', () => {
    const n1 = createNode({
      tags: { k: 'na', a: 'b' },
      id: 'n1',
    });
    const n2 = createNode({ id: 'n1', tags: { k: 'na', a: 'b' } });
    expect(n1).toBe(n2);
  });

  test('reuses already existing instance when lngLat', () => {
    const n1 = createNode({
      loc: genLngLat([1, 2]),
      tags: { k: 'na', a: 'b' },
      id: 'n1',
    });
    const n2 = createNode({
      id: 'n1',
      loc: genLngLat([1, 2]),
      tags: { k: 'na', a: 'b' },
    });

    expect(n1).toBe(n2);

    const n3 = createNode({
      loc: genLngLat([1, 2]),
      id: 'n1',
    });
    const n4 = createNode({
      id: 'n1',
      loc: genLngLat([1, 2]),
    });
    expect(n3).toBe(n4);
  });
});
