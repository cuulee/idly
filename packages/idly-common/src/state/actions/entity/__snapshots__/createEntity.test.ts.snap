// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create simple node simple test 1`] = `
Object {
  "attributes": Object {
    "changeset": undefined,
    "timestamp": undefined,
    "uid": undefined,
    "user": undefined,
    "version": undefined,
    "visible": undefined,
  },
  "id": "n1",
  "loc": Object {
    "lat": 0,
    "lon": 0,
  },
  "tags": Object {},
  "type": "node",
}
`;

exports[`create simple node simple test node with tags 1`] = `
Object {
  "attributes": Object {
    "changeset": undefined,
    "timestamp": undefined,
    "uid": undefined,
    "user": undefined,
    "version": undefined,
    "visible": undefined,
  },
  "id": "n1",
  "loc": Object {
    "lat": 0,
    "lon": 0,
  },
  "tags": Object {
    "k": "na",
  },
  "type": "node",
}
`;

exports[`create simple node works with full blown object 1`] = `
Object {
  "attributes": Object {
    "changeset": undefined,
    "timestamp": undefined,
    "uid": undefined,
    "user": undefined,
    "version": undefined,
    "visible": undefined,
  },
  "id": "n-a",
  "loc": Object {
    "lat": 2,
    "lon": 1,
  },
  "tags": Object {
    "k": "n-a4",
  },
  "type": "node",
}
`;
